### DISCUSSION TITLE
Compatibility with Eclipse 2024-12

### DISCUSSION THREAD

**Discussion by rombert** [Genuitec/Copilot4Eclipse-community#59] [Help] [answered]
Hi,

I tried to rebase my Eclipse installation to the upcoming 2024-12. I am using Oomph to define the features to be installed so I switched my existing 2024-09 update sites to 2024-12. The installation failed and I get the following root cause

```
[2024-12-03 14:29:48] Cannot complete the request.  Generating details.
[2024-12-03 14:29:48] ERROR: org.eclipse.equinox.p2.director code=10053 Cannot complete the install because one or more required items could not be found.
  at org.eclipse.oomph.util.OomphPlugin.coreException(OomphPlugin.java:296)
  at org.eclipse.oomph.p2.internal.core.ProfileTransactionImpl.resolve(ProfileTransactionImpl.java:446)
  at org.eclipse.oomph.p2.internal.core.ProfileTransactionImpl.commit(ProfileTransactionImpl.java:356)
  at org.eclipse.oomph.setup.p2.impl.P2TaskImpl.perform(P2TaskImpl.java:904)
  at org.eclipse.oomph.setup.internal.core.SetupTaskPerformer.doPerformNeededSetupTasks(SetupTaskPerformer.java:3864)
  at org.eclipse.oomph.setup.internal.core.SetupTaskPerformer.performNeededSetupTasks(SetupTaskPerformer.java:3792)
  at org.eclipse.oomph.setup.internal.core.SetupTaskPerformer.performTriggeredSetupTasks(SetupTaskPerformer.java:3773)
  at org.eclipse.oomph.setup.internal.core.SetupTaskPerformer.perform(SetupTaskPerformer.java:3651)
  at org.eclipse.oomph.setup.internal.installer.SimpleVariablePage.installPerform(SimpleVariablePage.java:1353)
  at org.eclipse.oomph.setup.internal.installer.SimpleVariablePage$20.run(SimpleVariablePage.java:1165)
  ERROR: org.eclipse.equinox.p2.director code=0 Software being installed: artificial_root 1.0.0.v1733232587972
  ERROR: org.eclipse.equinox.p2.director code=0 Missing requirement: Copilot4Eclipse 1.3.2.202411110916 (com.genuitec.copilot4eclipse 1.3.2.202411110916) requires 'osgi.bundle; org.eclipse.wildwebdeveloper.embedder.node [1.0.2,1.0.4)' but it could not be found
  ERROR: org.eclipse.equinox.p2.director code=1 Cannot satisfy dependency:
    ERROR: org.eclipse.equinox.p2.director code=0 From: artificial_root 1.0.0.v1733232587972
    ERROR: org.eclipse.equinox.p2.director code=0 To: org.eclipse.equinox.p2.iu; com.genuitec.copilot4eclipse.feature.feature.group 0.0.0
  ERROR: org.eclipse.equinox.p2.director code=1 Cannot satisfy dependency:
    ERROR: org.eclipse.equinox.p2.director code=0 From: Copilot4Eclipse Core 1.3.2.202411110916 (com.genuitec.copilot4eclipse.core.feature.feature.group 1.3.2.202411110916)
    ERROR: org.eclipse.equinox.p2.director code=0 To: org.eclipse.equinox.p2.iu; com.genuitec.copilot4eclipse [1.3.2.202411110916,1.3.2.202411110916]
  ERROR: org.eclipse.equinox.p2.director code=1 Cannot satisfy dependency:
    ERROR: org.eclipse.equinox.p2.director code=0 From: Copilot4Eclipse 1.3.2.202411110916 (com.genuitec.copilot4eclipse.feature.feature.group 1.3.2.202411110916)
    ERROR: org.eclipse.equinox.p2.director code=0 To: org.eclipse.equinox.p2.iu; com.genuitec.copilot4eclipse.core.feature.feature.group [1.3.2.202411110916,1.3.2.202411110916]

[2024-12-03 14:29:48] 
```

I pull in Wild Web Developer from http://download.eclipse.org/wildwebdeveloper/releases/latest and that update site provides org.eclipse.wildwebdeveloper.embedder.node.feature.feature.group with version 1.2.0.202411190713 .

Is this something that can be fixed/supported on the Copilot4Eclipse side?

**Comment by Bulletdev** [comment]
The error suggests a dependency compatibility issue between Copilot4Eclipse and Wild Web Developer. Specifically, Copilot4Eclipse is looking for org.eclipse.wildwebdeveloper.embedder.node in version range [1.0.2, 1.0.4), but the latest Wild Web Developer provides version 1.2.0.
This appears to be a version constraint problem that would need to be addressed by the Copilot4Eclipse plugin developers. They would need to:

Update their version range to include the newer Wild Web Developer version, or
Ensure compatibility with the latest Wild Web Developer release

I recommend:

Checking the Copilot4Eclipse GitHub issues or support channels
Reporting this compatibility issue to their development team
Temporarily using an older version of Wild Web Developer that matches the required version range

**Reply by rombert** [reply]
Thanks for the answer @Bulletdev . I did report the issue to the Copilot4Eclipse dev team, let's see what they have to say :-)

**Comment by Bulletdev** [comment]


**Reply by rombert** [reply]
But this is not answered, sorry.

**Comment by wayneparrott** [comment]
@rombert I'm looking into this issue. 

I would like to replicate your experience. Can you share any details about your update-site tweaks or cache changes?

Our team has been testing new Copilot4Eclipse installs and updates against the 2024-12 milestones with no issues even up to earlier today. Independently I just ran the following scenarios:
1. installed new 2024-09 oomph + Copilot4Eclipse from the Marketplace client. Then I added 2024-12 update-site, checked for updates and accepted all. No issues with Copilot4Eclipse observed.  
2. I then installed 2024-12.zip and Copilot4Eclipse from the Marketplace with no issues.
3. Installed new 2024-12 OOmph + Copilot4Eclipse from the Marketplace with no issues.

My interest in possible cache changes is that during Copilot4Eclipse installation it pulls its version of the embedder-node plugin from the 2024-03 site. Thus if the cache is goofed up such that this plugin is no longer present then that could explain things.

Wayne
Copilot4Eclipse Team

**Reply by rombert** [reply]
Thanks for looking into this @wayneparrott . I am using an Eclipse Oomph setup file (if that is the right term) as defined at https://wiki.eclipse.org/Eclipse_Oomph_Authoring . It may be that this kind of set up does not automatically follow associate p2 update sites. I will try adding a p2 update site for the older version of WWD to my set up and see what does. Will keep you posted.

**Reply by rombert** [reply]
The solution for me was to add https://download.eclipse.org/wildwebdeveloper/releases/1.3.3/ as an update site to my setup file, then the dependencies were resovled correctly. As your testing indicates, this was not an issue with the Copilot4Eclipse setup, but a limitation of Oomph (or incorrect usage on my behalf).

Thanks for helping me understanding the issue and finding a solution.

**Comment by brianvfernandes** [comment]
@rombert we have specifically added a dependency on `org.eclipse.wildwebdeveloper.embedder.node` in version range `[1.0.2, 1.0.4)` to support installations which still run using Java 17. Plugins in this range work with both Java 17 and Java 21, while later versions of the embedder.node plugin require Java 21. For some historical context, when WWD released plugin updates which required Java 21, a whole bunch of  Copilot4Eclipse installations on pre 2024-06 versions of Eclipse, broke (if they were running on Java 17).

**Note that Copilot4Eclipse must support a wide range of Eclipse releases, starting from 2023-06 (which could run with Java 17) to 2024-06 and above, which must run with Java 21.**

Coming to your error, here are some key notes:  

1. This is not an incompatibility or dependency conflict, simply an inability to find this specific version of the plugin. 
2. **All you have to do is add the `https://download.eclipse.org/releases/2024-03` update site to your list of Available Update sites (you are already familiar with the preference page), ensure it is enabled, and then reattempt the installation.** 
Note that the Copilot4Eclipse composite update site does add this reference to Eclipse automatically, so in most cases, a user does not need to do this and this just works, as @wayneparrott described above. However, given what you said about switching your update sites, I'm assuming you have this site disabled, or perhaps you have explicitly removed it.
3. The `org.eclipse.wildwebdeveloper.embedder.node` plugin is **not a singleton**. This means multiple versions can exist and be active in the same install. With Eclipse 2024-12 and the most recent WWD release, you will see both version `1.0.3` and I believe version `1.0.7` of this plugin in your installation.
4. The plugin will bind to the latest version of the embedded node _fragments_ in your installation. So even though we have version 1.0.3 referenced by Copilot4Eclipse, it will use the latest node binaries, shipped by the latest WWD version you have installed.

Hope you're able to move forward with step #2 - do let us know.

For some additional context - we like keeping Copilot4Eclipse compatible with as many releases of Eclipse as possible, this extends beyond standard Eclipse packages to third party IDEs, some of which are based on older releases of Eclipse which may not run with Java 21. If/when we drop this compatibility, this constraint will be removed, I hope this helps you understand the need for this specific dependency. Cheers!

**Reply by rombert** [reply]
Thanks for the detailed write-up @brianvfernandes . I fully understand the need to be compatible with multiple Eclipse release streams and Java versions. I think you landed on an elegant solution, but perhaps it's just my setup not being compatible with associate p2 update sites. I am using https://wiki.eclipse.org/Eclipse_Oomph_Authoring to define my Eclipse installations and I will try to add a WWD p2 update site that provides the needed, older, version of `org.eclipse.wildwebdeveloper.embedder.node`. Will keep you posted.

**Reply by rombert** [reply, marked as answer]
The solution for me was to add https://download.eclipse.org/wildwebdeveloper/releases/1.3.3/ as an update site to my setup file, then the dependencies were resovled correctly. As your explanation shows, this was not an issue with the Copilot4Eclipse setup, but a limitation of Oomph (or incorrect usage on my behalf).

Thanks for helping me understanding the issue and finding a solution.
